crates/bevy_ecs/src/query/access.rs

```diff
@@ -325,6 +375,38 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+#[derive(Clone, Eq, PartialEq)]
+struct AccessFilters<T> {
+    with: FixedBitSet,
+    without: FixedBitSet,
+    _index_type: PhantomData<T>,
+}
+
+impl<T: SparseSetIndex + fmt::Debug> fmt::Debug for AccessFilters<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AccessFilters")
+            .field("with", &FormattedBitSet::<T>::new(&self.with))
+            .field("without", &FormattedBitSet::<T>::new(&self.without))
+            .finish()
+    }
+}
+
+impl<T: SparseSetIndex> Default for AccessFilters<T> {
+    fn default() -> Self {
+        Self {
+            with: FixedBitSet::default(),
+            without: FixedBitSet::default(),
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T: SparseSetIndex> AccessFilters<T> {
+    fn is_ruled_out_by(&self, other: &Self) -> bool {
+        !self.with.is_disjoint(&other.without) || !self.without.is_disjoint(&other.with)
```

Comment:
Technically speaking to be complete this also needs `|| !self.with.is_disjoing(&self.without) || !other.with.is_disjoint(&other.without)` in order to handle extreme cases like `(With<T>, Without<T>)` being empty and thus disjoint with everything else. Not sure if we want to properly support those cases though, in practice if they happen they're almost always a mistake. The condition being at the end means it would be checked only if we're about to find a conflict, so the performance hit should not happen in the happy path.

However if it's kept like this I would add a comment explaining the missing conditions and why they're not implemented.

Reply:
Yeah, I was also thinking about this case. I'm inclined not to support it, I agree that it almost always means there's an error. Added a comment.